crates/red_knot_python_semantic/src/types.rs

```diff
@@ -399,7 +399,8 @@ pub enum Type<'db> {
     /// This variant should be created with the `todo_type!` macro.
     Todo(TodoType),
     /// The empty set of values
-    Never,
+    /// Synonym of `Never` is `NoReturn` and they both are represented by this type
+    Never(NeverType),
```

Comment:
To be honest I'm not sure if this will be useful for anything other than showing the type correctly to the user on hover. Without adding this field we show both `NoReturn` and `Never` as `Never`.
But this makes you type few more characters on pattern matching and when creating the type(The type creation can be simplified by making a function that would return the `Type::Never(NeverType::Never)`.)

Reply:
My feeling is that we do not need to maintain an internal distinction between `Never` and `NoReturn`, as types. (We should distinguish between them as known instances in the `typing` module, as you do.) They are the same type, and it's not even clear to me that it results in better UX to distinguish them. `NoReturn` is just an alternate form for function return annatations. The inferred return value from calling a function annotated to return `NoReturn` is IMO actually better revealed as `Never` than `NoReturn`, because the term `NoReturn` makes no sense outside the immediate return annotation context.

Reply:
> Let's make sure @AlexWaygood doesnt have a strong feeling otherwise, before you go through the work of removing this enum.

I agree! I had thoughts along the same lines, though I hadn't written them up yet. Mypy and pyright do the same, IIRC

Reply:
Thank you both. I reverted this one.